File: lib/src/transport/monitor_runner.dart
part of frugal;

/// MonitorRunner runs an FTransportMonitor when a transport is closed.
class MonitorRunner {
  final Logger log = new Logger('FTransportMonitor');
  FTransportMonitor _monitor;
  FTransport _transport;

  MonitorRunner(this._monitor, this._transport);

  Future onClose(Exception cause) async {
    if(cause == null) {
      _handleCleanClose();
    } else {
      await _handleUncleanClose(cause);
    }
  }

  void _handleCleanClose() {
    log.info('transport was closed cleanly');
    _monitor.onClosedCleanly();
  }

  Future _handleUncleanClose(Exception cause) async {
    log.warning('transport was closed uncleanly because: $cause');
    int wait = _monitor.onClosedUncleanly(cause);
    if(wait < 0) {
      log.warning('instructed not to repopen');
      return;
    }
    await _attemptReopen(wait);
  }

  Future _attemptReopen(int initialWait) async {
    int wait = initialWait;
    int prevAttempts = 0;

    while(wait >= 0) {
      log.info('attemptying to reopen after $wait ms');
      await new Future.delayed(new Duration(milliseconds: wait));

      try {
        await _transport.open();
      } on TTransportError catch (e) {
        log.warning('failed to reopen transport due to: $e');
        prevAttempts++;
        wait = _monitor.onReopenFailed(prevAttempts, wait);
        continue;
      }

      log.info('successfully reopened transport');
      _monitor.onReopenSucceeded();
    }

    log.warning('ReopenFailed callback instructed not to repoen, terminating');
  }
}

File: lib/java/src/main/java/com/workiva/frugal/transport/monitor/MonitorRunner.java
package com.workiva.frugal.transport.monitor;

import com.workiva.frugal.transport.FTransport;
import com.workiva.frugal.transport.FTransportClosedCallback;
import org.apache.thrift.transport.TTransportException;

import java.util.logging.Logger;

public class MonitorRunner implements FTransportClosedCallback {

    private static Logger LOGGER = Logger.getLogger(MonitorRunner.class.getName());

    private FTransportMonitor monitor;
    private FTransport transport;

    public MonitorRunner(FTransportMonitor monitor, FTransport transport) {
        this.monitor = monitor;
        this.transport = transport;
    }

    @Override
    public void onClose(Exception cause) {
        if (cause == null) {
            handleCleanClose();
        } else {
            handleUncleanClose(cause);
        }
    }

    private void handleCleanClose() {
        LOGGER.info("FTransport Monitor: FTransport was closed cleanly.");
        monitor.onClosedCleanly();
    }

    private void handleUncleanClose(Exception cause) {
        LOGGER.warning("FTransport Monitor: FTransport was closed uncleanly because: " + cause.getMessage());
        long wait = monitor.onClosedUncleanly(cause);
        if (wait < 0) {
            LOGGER.warning("FTransport Monitor: Instructed not to reopen.");
            return;
        }
        attemptReopen(wait);
    }

    private void attemptReopen(long initialWait) {
        long wait = initialWait;
        long prevAttempts = 0;

        while (wait >= 0) {
            LOGGER.info("FTransport Monitor: Attempting to reopen after " + wait + " ms");
            try {
                Thread.sleep(wait);
            } catch (InterruptedException e) {
                LOGGER.warning("FTransport Monitor: Reconnect wait interrupted: " + e.getMessage());
            }

            try {
                transport.open();
            } catch (TTransportException e) {
                LOGGER.warning("FTransport Monitor: Failed to reopen transport due to: " + e.getMessage());
                prevAttempts++;

                wait = monitor.onReopenFailed(prevAttempts, wait);
                continue;
            }

            LOGGER.info("FTransport Monitor: Successfully reopened!");
            monitor.onReopenSucceeded();
            return;
        }

        LOGGER.warning("FTransport Monitor: ReopenFailed callback instructed not to reopen. Terminating...");
    }

}
